File: supplier_management/supplier_management/report/supply_needs/supply_needs.py
# Copyright (c) 2013, Frappe Technologies Pvt. Ltd. and contributors
# For license information, please see license.txt

import frappe
from frappe import _
from frappe.utils import cint, flt


def execute(filters=None):
    columns = get_columns()
    data, total_count, selling_price_list = get_data(filters)
    message = {
        "total_count": total_count,
        "selling_price_list": selling_price_list,
        "filter_options": get_filter_options(),
    }
    return columns, data, message


def get_columns():
    columns = [
        {
            "label": _("Image"),
            "fieldname": "image",
            "fieldtype": "Data",
            "width": 100,
        },
        {
            "label": _("Item"),
            "fieldname": "item_code",
            "fieldtype": "Link",
            "options": "Item",
            "width": 120,
        },
        {
            "label": _("Item Name"),
            "fieldname": "item_name",
            "fieldtype": "Data",
            "width": 180,
        },
        {
            "label": _("Collection"),
            "fieldname": "custom_collection",
            "fieldtype": "Data",
            "width": 120,
        },
        {
            "label": _("Finition"),
            "fieldname": "custom_finition",
            "fieldtype": "Data",
            "width": 120,
        },
        {
            "label": _("Type Of Product"),
            "fieldname": "custom_type_de_produit",
            "fieldtype": "Data",
            "width": 140,
        },
        {
            "label": _("Brand"),
            "fieldname": "brand",
            "fieldtype": "Link",
            "options": "Brand",
            "width": 120,
        },
        {
            "label": _("Warehouse"),
            "fieldname": "warehouse",
            "fieldtype": "Data",
            "width": 120,
        },
        {
            "label": _("Supplier"),
            "fieldname": "supplier",
            "fieldtype": "Data",
            "width": 120,
        },
        {
            "label": _("Price"),
            "fieldname": "price",
            "fieldtype": "Currency",
            "options": "price_currency",
            "width": 120,
        },
        {
            "label": _("Available Qty"),
            "fieldname": "actual_qty",
            "fieldtype": "Data",
            "width": 150,
        },
        {
            "label": _("Backorder Qty"),
            "fieldname": "projected_qty",
            "fieldtype": "Data",
            "width": 220,
            "editable": 1,
        },
        {
            "label": _("Required Qty"),
            "fieldname": "reserved_qty",
            "fieldtype": "Data",
            "width": 150,
        },
        {
            "label": _("Requested Qty"),
            "fieldname": "indented_qty",
            "fieldtype": "Data",
            "width": 150,
        },
        {
            "label": _("Receivable Qty"),
            "fieldname": "ordered_qty",
            "fieldtype": "Data",
            "width": 150,
        },
    ]
    return columns


def get_data(filters):
    filters = filters or {}
    page_length = cint(filters.get("_page_length")) or 50
    page_start = max(cint(filters.get("_page_start")), 0)
    page_length = min(page_length, 500)
    selling_price_list = get_selling_price_list()

    conditions = ["item.disabled = 0", "item.is_stock_item = 1"]
    bin_conditions = []
    values = {"selling_price_list": selling_price_list}

    apply_filters(filters, conditions, bin_conditions, values)

    bin_where_clause = f"WHERE {' AND '.join(bin_conditions)}" if bin_conditions else ""
    where_clause = " AND ".join(conditions)
    values.update({"page_start": page_start, "page_length": page_length})

    sortable_fields = {
        "item_code": "item.name",
        "item_name": "item.item_name",
        "custom_collection": "item.custom_collection",
        "custom_finition": "item.custom_finition",
        "custom_type_de_produit": "item.custom_type_de_produit",
        "brand": "item.brand",
        "warehouse": "bin.warehouse",
        "supplier": "item_supplier.supplier",
        "price": "price.price_list_rate",
        "actual_qty": "bin.actual_qty",
        "projected_qty": "bin.projected_qty",
        "reserved_qty": "bin.reserved_qty",
        "indented_qty": "bin.indented_qty",
        "ordered_qty": "bin.ordered_qty",
    }
    sort_field = filters.get("_sort_field")
    sort_order = filters.get("_sort_order") if filters.get("_sort_order") in ("asc", "desc") else "asc"
    sort_expression = sortable_fields.get(sort_field)

    if sort_expression:
        order_by = f"{sort_expression} IS NULL ASC, {sort_expression} {sort_order.upper()}, item.name ASC"
    else:
        order_by = "bin.projected_qty IS NULL ASC, bin.projected_qty ASC, item.name ASC"

    total_count = frappe.db.sql(
        f"""
        SELECT COUNT(*)
        FROM `tabItem` item
        LEFT JOIN (
            SELECT
                item_code,
                CASE
                    WHEN COUNT(DISTINCT warehouse) = 1 THEN MIN(warehouse)
                    ELSE ''
                END AS warehouse,
                SUM(actual_qty) AS actual_qty,
                SUM(indented_qty) AS indented_qty,
                SUM(reserved_qty) AS reserved_qty,
                SUM(ordered_qty) AS ordered_qty,
                SUM(projected_qty) AS projected_qty
            FROM `tabBin`
            {bin_where_clause}
            GROUP BY item_code
        ) bin ON bin.item_code = item.name
        LEFT JOIN (
            SELECT
                parent,
                GROUP_CONCAT(DISTINCT supplier ORDER BY supplier SEPARATOR ', ') AS supplier
            FROM `tabItem Supplier`
            GROUP BY parent
        ) item_supplier ON item_supplier.parent = item.name
        LEFT JOIN `tabItem Price` price
            ON price.name = (
                SELECT item_price.name
                FROM `tabItem Price` item_price
                WHERE item_price.item_code = item.name
                    AND item_price.price_list = %(selling_price_list)s
                    AND item_price.selling = 1
                    AND COALESCE(item_price.customer, '') = ''
                    AND COALESCE(item_price.batch_no, '') = ''
                    AND (item_price.valid_from IS NULL OR item_price.valid_from <= CURDATE())
                    AND (item_price.valid_upto IS NULL OR item_price.valid_upto >= CURDATE())
                ORDER BY
                    CASE WHEN item_price.uom = item.stock_uom THEN 0 ELSE 1 END,
                    item_price.valid_from DESC,
                    item_price.modified DESC
                LIMIT 1
            )
        WHERE {where_clause}
        """,
        values,
    )[0][0]

    result = frappe.db.sql(
        f"""
        SELECT
            item.image,
            item.name AS item_code,
            item.item_name,
            item.custom_collection,
            item.custom_finition,
            item.custom_type_de_produit,
            item.brand,
            bin.warehouse,
            item_supplier.supplier,
            price.price_list_rate AS price,
            price.currency AS price_currency,
            bin.actual_qty,
            bin.indented_qty,
            bin.reserved_qty,
            bin.ordered_qty,
            bin.projected_qty
        FROM `tabItem` item
        LEFT JOIN (
            SELECT
                item_code,
                CASE
                    WHEN COUNT(DISTINCT warehouse) = 1 THEN MIN(warehouse)
                    ELSE ''
                END AS warehouse,
                SUM(actual_qty) AS actual_qty,
                SUM(indented_qty) AS indented_qty,
                SUM(reserved_qty) AS reserved_qty,
                SUM(ordered_qty) AS ordered_qty,
                SUM(projected_qty) AS projected_qty
            FROM `tabBin`
            {bin_where_clause}
            GROUP BY item_code
        ) bin ON bin.item_code = item.name
        LEFT JOIN (
            SELECT
                parent,
                GROUP_CONCAT(DISTINCT supplier ORDER BY supplier SEPARATOR ', ') AS supplier
            FROM `tabItem Supplier`
            GROUP BY parent
        ) item_supplier ON item_supplier.parent = item.name
        LEFT JOIN `tabItem Price` price
            ON price.name = (
                SELECT item_price.name
                FROM `tabItem Price` item_price
                WHERE item_price.item_code = item.name
                    AND item_price.price_list = %(selling_price_list)s
                    AND item_price.selling = 1
                    AND COALESCE(item_price.customer, '') = ''
                    AND COALESCE(item_price.batch_no, '') = ''
                    AND (item_price.valid_from IS NULL OR item_price.valid_from <= CURDATE())
                    AND (item_price.valid_upto IS NULL OR item_price.valid_upto >= CURDATE())
                ORDER BY
                    CASE WHEN item_price.uom = item.stock_uom THEN 0 ELSE 1 END,
                    item_price.valid_from DESC,
                    item_price.modified DESC
                LIMIT 1
            )
        WHERE {where_clause}
        ORDER BY
            {order_by}
        LIMIT %(page_length)s OFFSET %(page_start)s
        """,
        values,
        as_dict=1,
    )

    blank_if_missing_fields = (
        "item_name",
        "custom_collection",
        "custom_finition",
        "custom_type_de_produit",
        "brand",
        "warehouse",
        "supplier",
        "price",
        "price_currency",
        "actual_qty",
        "indented_qty",
        "reserved_qty",
        "ordered_qty",
        "projected_qty",
    )

    # Format image URLs for display and leave missing supply data empty.
    for row in result:
        if row.get("image"):
            row["image"] = f'<img src="{row["image"]}" style="max-width: 100px; max-height: 100px; object-fit: cover;" />'
        else:
            row["image"] = ""

        for fieldname in blank_if_missing_fields:
            if row.get(fieldname) is None:
                row[fieldname] = ""

    # Check for duplicate item_code within the same warehouse
    item_warehouse_map = {}
    for row in result:
        item_code = row.get("item_code")
        warehouse = row.get("warehouse") or "Unknown Warehouse"
        supplier = row.get("supplier") or "Unknown Supplier"

        if item_code:
            # Create a unique key combining item_code and warehouse
            key = (item_code, warehouse)
            if key in item_warehouse_map:
                item_warehouse_map[key].append(supplier)
            else:
                item_warehouse_map[key] = [supplier]

    # Find duplicates (same item_code in same warehouse with multiple suppliers)
    duplicates = {}
    for (item_code, warehouse), suppliers in item_warehouse_map.items():
        if len(suppliers) > 1:
            if item_code not in duplicates:
                duplicates[item_code] = {}
            duplicates[item_code][warehouse] = suppliers

    if duplicates:
        # Prepare the message content
        message_parts = []
        for item_code, warehouses in duplicates.items():
            for warehouse, suppliers in warehouses.items():
                message_parts.append(
                    f"{item_code} (Warehouse: {warehouse}) => {', '.join(suppliers)}"
                )

        message = "<br>".join(message_parts)

        # Show the popup message
        frappe.msgprint(
            title=_("Duplicate Item Codes"),
            msg=_(
                "The following item codes have multiple suppliers in the same warehouse:<br>{0}"
            ).format(message),
            indicator="orange",
        )

    return result, total_count, selling_price_list


def apply_filters(filters, conditions, bin_conditions, values):
    text_filters = {
        "item_code": "item.name",
        "item_name": "item.item_name",
        "custom_collection": "item.custom_collection",
        "custom_finition": "item.custom_finition",
        "custom_type_de_produit": "item.custom_type_de_produit",
        "brand": "item.brand",
        "supplier": "item_supplier.supplier",
    }
    qty_filters = {
        "price": "price.price_list_rate",
        "actual_qty": "bin.actual_qty",
        "projected_qty": "bin.projected_qty",
        "reserved_qty": "bin.reserved_qty",
        "indented_qty": "bin.indented_qty",
        "ordered_qty": "bin.ordered_qty",
    }

    for fieldname, expression in text_filters.items():
        if fieldname in filters and filters.get(fieldname) not in (None, ""):
            add_text_filter(conditions, values, expression, fieldname, filters.get(fieldname))

    if filters.get("warehouse"):
        add_text_filter(bin_conditions, values, "warehouse", "warehouse", filters.get("warehouse"))
        conditions.append("bin.item_code IS NOT NULL")

    for fieldname, expression in qty_filters.items():
        if fieldname in filters and filters.get(fieldname) not in (None, ""):
            add_numeric_filter(conditions, values, expression, fieldname, filters.get(fieldname))


def add_text_filter(conditions, values, expression, fieldname, raw_value):
    value = str(raw_value).strip()
    if not value:
        return

    if value.startswith("!="):
        key = make_filter_key(values, fieldname)
        conditions.append(f"COALESCE({expression}, '') NOT LIKE %({key})s ESCAPE '\\\\'")
        values[key] = f"%{escape_like(value[2:].strip())}%"
        return

    if value.startswith("="):
        key = make_filter_key(values, fieldname)
        conditions.append(f"COALESCE({expression}, '') = %({key})s")
        values[key] = value[1:].strip()
        return

    key = make_filter_key(values, fieldname)
    conditions.append(f"COALESCE({expression}, '') LIKE %({key})s ESCAPE '\\\\'")
    values[key] = f"%{escape_like(value)}%"


def add_numeric_filter(conditions, values, expression, fieldname, raw_value):
    value = str(raw_value).strip()
    if not value:
        return

    if ":" in value and not value.startswith(("=", "!", "<", ">")):
        start, end = value.split(":", 1)
        start_key = make_filter_key(values, f"{fieldname}_start")
        end_key = make_filter_key(values, f"{fieldname}_end")
        conditions.append(f"COALESCE({expression}, 0) BETWEEN %({start_key})s AND %({end_key})s")
        values[start_key] = flt(start)
        values[end_key] = flt(end)
        return

    operators = (">=", "<=", "!=", ">", "<", "=")
    operator = "="
    number = value

    for possible_operator in operators:
        if value.startswith(possible_operator):
            operator = possible_operator
            number = value[len(possible_operator) :].strip()
            break

    key = make_filter_key(values, fieldname)
    conditions.append(f"COALESCE({expression}, 0) {operator} %({key})s")
    values[key] = flt(number)


def make_filter_key(values, fieldname):
    key = fieldname
    index = 1

    while key in values:
        key = f"{fieldname}_{index}"
        index += 1

    return key


def get_selling_price_list():
    return frappe.db.get_single_value("Selling Settings", "selling_price_list") or frappe.db.get_value(
        "Price List",
        {"selling": 1, "enabled": 1},
        "name",
        order_by="name",
    )


def get_filter_options():
    return {
        "custom_collection": get_item_field_options("custom_collection"),
        "custom_finition": get_item_field_options("custom_finition"),
        "custom_type_de_produit": get_item_field_options("custom_type_de_produit"),
        "brand": get_brand_options(),
        "warehouse": get_warehouse_options(),
        "supplier": get_supplier_options(),
    }


def get_item_field_options(fieldname):
    if fieldname not in {
        "custom_collection",
        "custom_finition",
        "custom_type_de_produit",
    }:
        return []

    rows = frappe.db.sql(
        f"""
        SELECT DISTINCT {fieldname} AS value
        FROM `tabItem`
        WHERE disabled = 0
            AND is_stock_item = 1
            AND COALESCE({fieldname}, '') != ''
        ORDER BY {fieldname}
        """,
        as_dict=1,
    )

    return make_options(rows)


def get_brand_options():
    rows = frappe.db.sql(
        """
        SELECT DISTINCT
            item.brand AS value,
            COALESCE(brand.brand, item.brand) AS label
        FROM `tabItem` item
        LEFT JOIN `tabBrand` brand
            ON brand.name = item.brand
        WHERE item.disabled = 0
            AND item.is_stock_item = 1
            AND COALESCE(item.brand, '') != ''
        ORDER BY label
        """,
        as_dict=1,
    )

    return make_options(rows)


def get_warehouse_options():
    rows = frappe.db.sql(
        """
        SELECT DISTINCT
            bin.warehouse AS value,
            COALESCE(warehouse.warehouse_name, bin.warehouse) AS label
        FROM `tabBin` bin
        INNER JOIN `tabItem` item
            ON item.name = bin.item_code
            AND item.disabled = 0
            AND item.is_stock_item = 1
        LEFT JOIN `tabWarehouse` warehouse
            ON warehouse.name = bin.warehouse
        WHERE COALESCE(bin.warehouse, '') != ''
        ORDER BY label
        """,
        as_dict=1,
    )

    return make_options(rows)


def get_supplier_options():
    rows = frappe.db.sql(
        """
        SELECT DISTINCT
            item_supplier.supplier AS value,
            COALESCE(supplier.supplier_name, item_supplier.supplier) AS label
        FROM `tabItem Supplier` item_supplier
        INNER JOIN `tabItem` item
            ON item.name = item_supplier.parent
            AND item.disabled = 0
            AND item.is_stock_item = 1
        LEFT JOIN `tabSupplier` supplier
            ON supplier.name = item_supplier.supplier
        WHERE COALESCE(item_supplier.supplier, '') != ''
        ORDER BY label
        """,
        as_dict=1,
    )

    return make_options(rows)


def make_options(rows):
    return [
        {
            "value": row.value,
            "label": row.get("label") or row.value,
        }
        for row in rows
        if row.value
    ]


def escape_like(value):
    return str(value).replace("\\", "\\\\").replace("%", "\\%").replace("_", "\\_")

File: supplier_management/supplier_management/report/supply_needs/supply_needs.js
// Copyright (c) 2013, Frappe Technologies Pvt. Ltd. and contributors
// For license information, please see license.txt

frappe.query_reports["Supply Needs"] = {
	filters: [
		{
			fieldname: "item_code",
			label: __("Item"),
			fieldtype: "Link",
			options: "Item",
		},
		{
			fieldname: "warehouse",
			label: __("Warehouse"),
			fieldtype: "Link",
			options: "Warehouse",
		},
		{
			fieldname: "supplier",
			label: __("Supplier"),
			fieldtype: "Link",
			options: "Supplier",
		},
	],

	page_length: 50,
	current_page: 1,
	total_count: 0,
	sort_field: null,
	sort_order: "none",

	projectedQtyIndex: 13,
	formatter(value, row, column, data, default_formatter) {
		value = default_formatter(value, row, column, data);

		// Highlight negative projected qty in red
		if (column.fieldname === "projected_qty" && data.projected_qty < 0) {
			value = `<span style="color:red; font-weight: bold;">${value}</span>`;
		}

		return value;
	},

	get_datatable_options(options) {
		return Object.assign(options, {
			checkboxColumn: true,
			cellHeight: 64,
			dynamicRowHeight: false,
		});
	},

	after_datatable_render(datatable) {
		this.patch_datatable_sorting(datatable);
		this.disable_image_column_controls(datatable);
		this.update_pagination();
		this.set_table_height(datatable);
	},

	onload(report) {
		this.report = report;
		this.inject_styles();
		this.patch_report_pagination(report);

		// Material Request Button - Using exact Sales Order pattern
		report.page.add_inner_button(__("Material Request"), () => {
			this.make_material_request(report);
		}, __("Create"));

		// Purchase Order Button - Using exact Sales Order pattern  
		report.page.add_inner_button(__("Purchase Order"), () => {
			this.make_purchase_order(report);
		}, __("Create"));

		// Set Create as primary button group
		report.page.set_inner_btn_group_as_primary(__("Create"));
	},

	inject_styles() {
		if (document.getElementById("supply-needs-report-styles")) return;

		const style = document.createElement("style");
		style.id = "supply-needs-report-styles";
		style.textContent = `
			[data-route="query-report/Supply Needs"] .supply-needs-pagination {
				display: flex;
				align-items: center;
				justify-content: space-between;
				gap: 16px;
				width: 100%;
				min-height: 36px;
				margin-bottom: 8px;
			}
			[data-route="query-report/Supply Needs"] .supply-needs-row-count {
				display: inline-flex;
				align-items: center;
				white-space: nowrap;
				font-weight: 500;
			}
			[data-route="query-report/Supply Needs"] .supply-needs-page-size,
			[data-route="query-report/Supply Needs"] .supply-needs-page-buttons {
				display: inline-flex;
				align-items: center;
				gap: 6px;
			}
			[data-route="query-report/Supply Needs"] .supply-needs-page-size select {
				width: 72px;
				height: 26px;
				padding-top: 2px;
				padding-bottom: 2px;
			}
			[data-route="query-report/Supply Needs"] .supply-needs-page-status {
				min-width: 170px;
				text-align: center;
			}
			[data-route="query-report/Supply Needs"] .datatable {
				border: 1px solid var(--border-color);
				border-radius: 6px;
				overflow: hidden;
			}
			[data-route="query-report/Supply Needs"] .datatable .dt-scrollable {
				min-height: 420px;
				max-height: calc(100vh - 260px);
			}
			[data-route="query-report/Supply Needs"] .datatable .dt-cell__content img {
				max-width: 56px !important;
				max-height: 56px !important;
				object-fit: cover;
				border-radius: 4px;
			}
			[data-route="query-report/Supply Needs"] .datatable .dt-header {
				background: var(--subtle-accent);
			}
			[data-route="query-report/Supply Needs"] .datatable .dt-cell--header .dt-cell__content {
				font-weight: 600;
			}
			[data-route="query-report/Supply Needs"] .datatable .dt-row-filter .dt-cell {
				height: 35px;
			}
			[data-route="query-report/Supply Needs"] .datatable .dt-filter:disabled {
				background: var(--disabled-control-bg);
				cursor: not-allowed;
			}
		`;
		document.head.appendChild(style);
	},

	patch_report_pagination(report) {
		if (report.supply_needs_pagination_patched) return;

		report.supply_needs_pagination_patched = true;
		const original_get_filter_values = report.get_filter_values.bind(report);
		const original_prepare_report_data = report.prepare_report_data.bind(report);
		const original_show_footer_message = report.show_footer_message.bind(report);

		report.get_filter_values = (...args) => {
			const filters = original_get_filter_values(...args) || {};
			const report_filter_signature = JSON.stringify(filters);

			if (
				this.last_report_filter_signature !== undefined &&
				this.last_report_filter_signature !== report_filter_signature
			) {
				this.current_page = 1;
			}

			this.last_report_filter_signature = report_filter_signature;
			filters._page_length = this.page_length;
			filters._page_start = (this.current_page - 1) * this.page_length;

			if (this.sort_field && this.sort_order !== "none") {
				filters._sort_field = this.sort_field;
				filters._sort_order = this.sort_order;
			}

			return filters;
		};

		report.prepare_report_data = (data) => {
			original_prepare_report_data(data);
			this.prepare_columns();
			this.total_count = data.message?.total_count || report.data.length;
			const last_page = this.get_last_page();
			if (this.current_page > last_page) {
				this.current_page = last_page;
				setTimeout(() => this.report.refresh(), 0);
			}
			this.apply_column_sort_state();
		};

		report.show_footer_message = () => {
			original_show_footer_message();
			this.render_pagination_controls();
			this.update_pagination();
		};
	},

	make_pagination_controls() {
		if (this.pagination_wrapper) return;

		this.pagination_wrapper = $(`
			<div class="supply-needs-pagination">
				<span class="supply-needs-row-count text-muted"></span>
				<div class="supply-needs-page-size">
					<span class="text-muted">${__("Rows")}</span>
					<select class="form-control input-xs" data-action="page_length">
						<option value="25">25</option>
						<option value="50" selected>50</option>
						<option value="100">100</option>
						<option value="250">250</option>
					</select>
				</div>
				<div class="supply-needs-page-buttons">
					<button class="btn btn-xs btn-default" data-action="first">${__("First")}</button>
					<button class="btn btn-xs btn-default" data-action="prev">${__("Previous")}</button>
					<span class="supply-needs-page-status text-muted"></span>
					<button class="btn btn-xs btn-default" data-action="next">${__("Next")}</button>
					<button class="btn btn-xs btn-default" data-action="last">${__("Last")}</button>
				</div>
			</div>
		`);

		this.pagination_wrapper.on("change", '[data-action="page_length"]', (event) => {
			this.page_length = cint(event.currentTarget.value) || 50;
			this.current_page = 1;
			this.render_current_page();
		});

		this.pagination_wrapper.on("click", "button[data-action]", (event) => {
			const action = event.currentTarget.dataset.action;
			const last_page = this.get_last_page();

			if (action === "first") this.current_page = 1;
			if (action === "prev") this.current_page = Math.max(1, this.current_page - 1);
			if (action === "next") this.current_page = Math.min(last_page, this.current_page + 1);
			if (action === "last") this.current_page = last_page;

			this.render_current_page();
		});
	},

	render_pagination_controls() {
		this.make_pagination_controls();
		const footer = this.report?.$report_footer;

		if (!footer?.length) return;

		footer.find(".supply-needs-pagination").remove();
		footer.prepend(this.pagination_wrapper);
	},

	update_pagination() {
		if (!this.pagination_wrapper) return;

		const total = this.total_count;
		const page_length = this.page_length;
		const last_page = this.get_last_page();
		const start = total ? (this.current_page - 1) * page_length + 1 : 0;
		const end = total ? Math.min(this.current_page * page_length, total) : 0;

		this.pagination_wrapper
			.find(".supply-needs-row-count")
			.text(__("{0} products loaded", [total]));
		this.pagination_wrapper
			.find(".supply-needs-page-status")
			.text(__("Showing {0}-{1} of {2}", [start, end, total]));
		this.pagination_wrapper
			.find('[data-action="page_length"]')
			.val(this.page_length);

		this.pagination_wrapper.find('[data-action="first"], [data-action="prev"]').prop("disabled", this.current_page <= 1);
		this.pagination_wrapper.find('[data-action="next"], [data-action="last"]').prop("disabled", this.current_page >= last_page);
	},

	prepare_columns() {
		if (!this.report?.columns) return;

		const image_column = this.report.columns.find((column) => {
			const fieldname = column.fieldname || column.id;
			return fieldname === "image";
		});
		if (!image_column) return;

		image_column.sortable = false;
		image_column.dropdown = false;
		image_column.editable = false;
		image_column.focusable = false;
	},

	render_current_page() {
		this.current_page = Math.min(this.current_page, this.get_last_page());
		this.report.refresh();
	},

	get_last_page() {
		return Math.max(1, Math.ceil(this.total_count / this.page_length));
	},

	patch_datatable_sorting(datatable) {
		if (!datatable || datatable.supply_needs_sort_patched) return;

		datatable.supply_needs_sort_patched = true;
		datatable.sortColumn = (colIndex, sortOrder) => {
			const column = datatable.datamanager.getColumn(colIndex);
			const fieldname = column?.fieldname || column?.id;

			if (!fieldname || ["_checkbox", "_rowIndex"].includes(fieldname)) return;

			this.sort_field = fieldname;
			this.sort_order = sortOrder || "none";
			this.current_page = 1;
			this.render_current_page();
		};
	},

	disable_image_column_controls(datatable) {
		const image_column = datatable?.datamanager
			?.getColumns()
			.find((column) => (column.fieldname || column.id) === "image");

		if (!image_column) return;

		const filter = datatable.header.querySelector(
			`.dt-cell--col-${image_column.colIndex} .dt-filter`
		);

		if (filter) {
			filter.disabled = true;
			filter.placeholder = "";
			filter.title = __("Image column is not searchable");
		}
	},

	apply_column_sort_state() {
		if (!this.report?.columns) return;

		this.report.columns.forEach((column) => {
			const fieldname = column.fieldname || column.id;
			column.sortOrder = fieldname === this.sort_field ? this.sort_order : "none";
		});
	},

	set_table_height(datatable) {
		const scrollable = datatable?.bodyScrollable;
		if (!scrollable) return;

		scrollable.style.height = "calc(100vh - 260px)";
		scrollable.style.maxHeight = "calc(100vh - 260px)";
		scrollable.style.minHeight = "420px";
		scrollable.style.overflowY = "auto";
	},

	parseFormattedNumber(value, numberFormat) {
		if (typeof value !== "string" || !value.trim()) {
			return NaN;
		}

		// Infer separators from number format
		let thousands_sep = numberFormat.includes(' ') ? ' ' : ',';
		let decimal_sep = numberFormat.includes(',') ? ',' : '.';

		// Replace thousands separator (remove it)
		value = value.replace(new RegExp('\\' + thousands_sep, 'g'), '');

		// Replace decimal separator with dot
		if (decimal_sep !== '.') {
			value = value.replace(decimal_sep, '.');
		}

		// Convert to float
		return parseFloat(value);
	},
	make_material_request(report) {
		let selected_rows = [];
		const numberFormat = get_number_format();

		// Get selected rows from the datatable
		if (frappe.query_report.datatable) {
			const checked_rows = frappe.query_report.datatable.rowmanager.getCheckedRows();
			selected_rows = checked_rows.map(i => {
				let rowData = frappe.query_report.data[i];
				let domRow = frappe.query_report.datatable.rowmanager.getRow$(i);
				let projectedQtyText = jQuery(domRow).find(`[data-col-index=${this.projectedQtyIndex}]`).text().trim();
				rowData.projected_qty = this.parseFormattedNumber(projectedQtyText, numberFormat);
				return rowData;
			});
		}
		// Validate selected rows
		if (!selected_rows.length) {
			frappe.throw({
				message: __("Please select rows to create Material Request"),
				title: __("No Items Selected"),
				indicator: "blue",
			});
		}

		// Filter valid rows
		const valid_rows = selected_rows.filter(row => {
			if (!row.item_code) return false;
			if (!row.projected_qty || isNaN(row.projected_qty)) return false;
			return true;
		});

		if (!valid_rows.length) {
			frappe.throw(__("No valid items to create Material Request."));
		}

		// Create a new Material Request document
		frappe.model.with_doctype("Material Request", () => {
			const doc = frappe.model.get_new_doc("Material Request");
			doc.material_request_type = "Purchase";
			doc.schedule_date = frappe.datetime.add_days(frappe.datetime.get_today(), 7);

			// Add valid rows as child items
			valid_rows.forEach(row => {
				const item = frappe.model.add_child(doc, "items");
				item.item_code = row.item_code;
				item.qty = Math.abs(row.projected_qty);
				item.warehouse = row.warehouse;
				item.schedule_date = frappe.datetime.add_days(frappe.datetime.get_today(), 7);

				// Fetch additional item details to auto-populate UOM
				frappe.call({
					method: "erpnext.stock.get_item_details.get_item_details",
					args: {
						args: {
							item_code: item.item_code,
							warehouse: item.warehouse,
							doctype: doc.doctype,
							company: doc.company,
							qty: item.qty,
						},
					},
					callback: function (r) {
						if (!r.exc && r.message) {
							item.uom = r.message.uom; // Set UOM from item master
							item.stock_uom = r.message.stock_uom; // Set stock UOM
							item.conversion_factor = r.message.conversion_factor; // Set conversion factor
							item.item_name = r.message.item_name; // Set item name
							refresh_field("items");
						}
					},
				});
			});

			// Navigate to the new Material Request form
			frappe.set_route("Form", "Material Request", doc.name).then(() => {
				// Refresh the form to show the added items
				cur_frm.refresh_fields();
			});
		});
	},

	make_purchase_order(report) {
		// Get selected rows
		let selected_rows = [];
		const numberFormat = get_number_format();

		if (frappe.query_report.datatable) {
			const checked_rows = frappe.query_report.datatable.rowmanager.getCheckedRows();
			selected_rows = checked_rows.map(i => {
				let rowData = frappe.query_report.data[i];
				let domRow = frappe.query_report.datatable.rowmanager.getRow$(i);
				let projectedQtyText = jQuery(domRow).find(`[data-col-index=${this.projectedQtyIndex}]`).text().trim();
				rowData.projected_qty = this.parseFormattedNumber(projectedQtyText, numberFormat);
				return rowData;
			});
		}

		if (!selected_rows.length) {
			frappe.throw({
				message: __("Please select rows to create Purchase Order"),
				title: __("No Items Selected"),
				indicator: "blue",
			});
		}

		// Filter valid rows
		const valid_rows = selected_rows.filter((row) => {
			if (!row.item_code) return false;
			if (!row.projected_qty || isNaN(row.projected_qty)) return false;
			return true;
		});

		if (!valid_rows.length) {
			frappe.throw(__("No valid items to create Purchase Order."));
		}

		// Create a new Purchase Order document
		frappe.model.with_doctype("Purchase Order", () => {
			const doc = frappe.model.get_new_doc("Purchase Order");

			// Add valid rows as child items
			valid_rows.forEach((row) => {
				const item = frappe.model.add_child(doc, "items");
				item.item_code = row.item_code;
				item.qty = Math.abs(row.projected_qty);
				item.warehouse = row.warehouse;
				item.schedule_date = frappe.datetime.add_days(frappe.datetime.get_today(), 7);

				// Fetch additional item details to auto-populate UOM and other fields
				frappe.call({
					method: "erpnext.stock.get_item_details.get_item_details",
					args: {
						args: {
							item_code: item.item_code,
							warehouse: item.warehouse,
							doctype: doc.doctype,
							company: doc.company,
							qty: item.qty,
						},
					},
					callback: function (r) {
						if (!r.exc && r.message) {
							item.uom = r.message.uom; // Set UOM from item master
							item.stock_uom = r.message.stock_uom; // Set stock UOM
							item.conversion_factor = r.message.conversion_factor; // Set conversion factor
							item.item_name = r.message.item_name; // Set item name
							refresh_field("items");
						}
					},
				});
			});

			// Navigate to the new Purchase Order form
			frappe.set_route("Form", "Purchase Order", doc.name).then(() => {
				// Refresh the form to show the added items
				cur_frm.refresh_fields();
			});
		});
	},
};
